### PR TITLE
open datasets

### PR DESCRIPTION
**Pull request by NicolasLegendre1**
The file dataset.py is usefull for VAE. It contains functions to open database and also to control VAE architecture

### CODE REVIEW

**Review by ninamiolane — CHANGES_REQUESTED** [review]
Thank you for this first PR!! 🎉 Great code. A few comments:
- I think functions related to opening datasets should go to the repo `ioSPI`, @fredericpoitevin what do you think? see [here](https://github.com/compSPI/compSPI): "ioSPI: Methods and tools to read and write data in all formats. Also provides methods to use other SPI tools developed outside of compSPI."
- IMPORTANT: you need to add unit-test functions to each function that you add in a PR (it *cannot* be in another PR), see the [contributing guidelines](https://github.com/compspi/compspi/blob/master/docs/contributing.rst) here: "Submit your code with associated unit tests.". For this, add a folder called `tests` at the same level of `models` and add a module `test_dataset.py` (needs to have the same name than the module.py containing the functions being tested). Then, you add tests [using the syntax here](https://github.com/geomstats/geomstats/blob/master/tests/tests_geomstats/test_backends.py) for example, but using `unittest.TestCase` instead of `geomstats.tests.TestCase`.
- Great docstrings! They are missing some elements though, for example the oneliner descriptor. See our [contributing guidelines](https://github.com/compspi/compspi/blob/master/docs/contributing.rst#writing-docstrings) for an example.
I'll put my other comments line by line.

File: models/dataset.py
Comment (ninamiolane) [inline models/dataset.py:2]: The docstring at the top of the file should also briefly describe what is in the file, see here for example:
https://github.com/geomstats/geomstats/blob/471a70578afe3926520a7fd8247e73ca05b387f5/tests/tests_geomstats/test_backends.py#L1

File: models/dataset.py
Comment (ninamiolane) [inline models/dataset.py:7]: Imports order:
https://www.python.org/dev/peps/pep-0008/#imports

This is not too bad though, as we have bots on this repo that should correct the order of your imports automatically.

File: models/dataset.py
Comment (ninamiolane) [inline models/dataset.py:17]: This is not PEP8 compliant: PEP8 recommends spaces around operators, in this case around the `+`.

To see all of your PEP8 errors, download `black`, pip install black, and use:
`black -S dataset.py`: it will reformat your code for it to be PEP8 compliant.

File: models/dataset.py
Comment (ninamiolane) [inline models/dataset.py:25]: parameters' names cannot be in uppercase: python thinks it is a constant (see, `D3` has a different color), please change to lowercase letters. something like `is_2d` would also be clearer.

File: models/dataset.py
Comment (ninamiolane) [inline models/dataset.py:29]: Docstring is almost perfect! Not quite though, please look at the contributing guidelines to finish to clean it

File: models/dataset.py
Comment (ninamiolane) [inline models/dataset.py:40]: `path.lower().endswith(".h5")` would be cleaner

File: models/dataset.py
Comment (ninamiolane) [inline models/dataset.py:47]: no uppercase if not a constant. if you use uppercase letters, python assumes it is either a constant variable or a Python class.

Also, `n_imgs` would be clearer as a name

File: models/dataset.py
Comment (ninamiolane) [inline models/dataset.py:53]: img_shape.ndim

File: models/dataset.py
Comment (ninamiolane) [inline models/dataset.py:48]: Ditto

File: models/dataset.py
Comment (ninamiolane) [inline models/dataset.py:83]: you mean `==`?

File: models/dataset.py
Comment (ninamiolane) [inline models/dataset.py:105]: n_imgs

File: models/dataset.py
Comment (ninamiolane) [inline models/dataset.py:106]: n_val

Comment (ninamiolane) [reply models/dataset.py:106]: (try to use variable names that will help the collaborators to understand your code)

File: models/dataset.py
Comment (ninamiolane) [inline models/dataset.py:115]: Constants variables should go at the top of the file.

File: models/dataset.py
Comment (ninamiolane) [inline models/dataset.py:116]: We should decide on using ' or " for quotes everywhere. I recommend "

Comment (ninamiolane) [reply models/dataset.py:116]: `simulated_2D`: in order to be consistent with the other namings below

File: models/dataset.py
Comment (ninamiolane) [inline models/dataset.py:122]: no camelcase, let's rename this .npy with _ and lowercase letters

File: models/dataset.py
Comment (ninamiolane) [inline models/dataset.py:133]: no camelcase in python

File: models/dataset.py
Comment (ninamiolane) [inline models/dataset.py:137]: English: CONSTANTS

File: models/dataset.py
Comment (ninamiolane) [inline models/dataset.py:170]: Since we have a lot of CONSTANTS here, let's organize them in yml files:
`datasets_config.yml` and `nn_config.yml` and import the yml file as a dictionary: https://stackabuse.com/reading-and-writing-yaml-to-a-file-in-python

File: models/dataset.py
Comment (ninamiolane) [inline models/dataset.py:174]: No camelcase: rename

File: models/dataset.py
Comment (ninamiolane) [inline models/dataset.py:151]: no uppercase letters in the keys of the dict: rename

**Comment by ninamiolane** [comment]
I've put a lot of comments, but that's normal: it's always like this for a first PR! Actually, it's not that many comments for a first PR, so congratulations for this clean code!

@fredericpoitevin we should put the GitHub actions with the unit-tests soon-ish, to make sure that we do not add un-tested code to the repos, otherwise the problems will quickly blow up

**Comment by fredericpoitevin** [comment]
> Thank you for this first PR!! 🎉 Great code. A few comments:
> 
> * I think functions related to opening datasets should go to the repo `ioSPI`, @fredericpoitevin what do you think? see [here](https://github.com/compSPI/compSPI): "ioSPI: Methods and tools to read and write data in all formats. Also provides methods to use other SPI tools developed outside of compSPI."
> * IMPORTANT: you need to add unit-test functions to each function that you add in a PR (it _cannot_ be in another PR), see the [contributing guidelines](https://github.com/compspi/compspi/blob/master/docs/contributing.rst) here: "Submit your code with associated unit tests.". For this, add a folder called `tests` at the same level of `models` and add a module `test_dataset.py` (needs to have the same name than the module.py containing the functions being tested). Then, you add tests [using the syntax here](https://github.com/geomstats/geomstats/blob/master/tests/tests_geomstats/test_backends.py) for example, but using `unittest.TestCase` instead of `geomstats.tests.TestCase`.
> * Great docstrings! They are missing some elements though, for example the oneliner descriptor. See our [contributing guidelines](https://github.com/compspi/compspi/blob/master/docs/contributing.rst#writing-docstrings) for an example.
>   I'll put my other comments line by line.

Thanks @ninamiolane for the detailed review!
I agree, opening datasets belong to https://github.com/compSPI/ioSPI.
@NicolasLegendre1 let me create a `tests` directory in ioSPI with an example, then please follow Nina's guidelines and comments.

File: models/dataset.py
Comment (NicolasLegendre1) [reply models/dataset.py:17]: thank you I'm already using black on spyder,

File: models/dataset.py
Comment (NicolasLegendre1) [reply models/dataset.py]: yes

File: models/dataset.py
Comment (ninamiolane) [reply models/dataset.py]: Hmm that's weird that it didn't catch the missing spaces around the operator + 🤔

File: models/dataset.py
Comment (NicolasLegendre1) [reply models/dataset.py]: i think it's good to have acces to this constants, so you  are able to change them really quickly

File: models/dataset.py
Comment (fredericpoitevin) [reply models/dataset.py:124]: Agreed. Also, I think Nina suggested storing this in a `.json` file, I think this is a good idea. 

@NicolasLegendre1 here is how you could do it:

# 1. Create a `dataset.json` file with the following content:
```json
{
  "PATHS": [
    {
      "simulated_2d": "cryo_sim_128x128.npy",
      "real_2d30": class2D_30_sort.h5",
      "real_2d39": "class2D_39_sort.h5",
      "real_2d93": "class2D_93_sort.h5",
      "refine_3D": "refine3D_180x180_sort.h5",
      "class_3d9090": "class3D_90x90_sort.h5",
      "simulated_3d": "concat_simulated.npy",
      "simulated_3d_noise": "cryo_sim_128x128.npy",
      "real_3d": "data.hdf5",
      "4points": "4points.npy",
      "4points1": "4points1.npy",
      "4points_3d": "3d_images.npy"
    }
  ],
  "SHAPES": [
    {
      "simulated_3d": (1, 320, 320),
      "simulated_2d": (1, 128, 128),
      "4points": (1, 128, 128),
      "4points1": (1, 64, 64),
      "4points_3d": (64, 64, 64)
    }
  ],
  etc...
}
```

# 2. add a function that loads it in `dataset.py`:
```python
import json

def load_parameters(path):
  with open(path) as json_file:
    parameters = json.load(json_file)
```

you can then get the values doing this, for example:
`filename = parameters['PATHS']['simulated_2d']`

File: models/dataset.py
Comment (fredericpoitevin) [inline models/dataset.py:17]: Is this absolutely needed?

File: models/nn.py
Comment (fredericpoitevin) [inline models/nn.py:1]: Exciting!!!

**Review by fredericpoitevin — CHANGES_REQUESTED** [review]
This is very exciting, thank you @NicolasLegendre1 !

We will probably move most of `dataset.py` to ioSPI soon, but in the meantime, please address the few comments I made.

For `nn.py`, I think we want to implement unit tests before merging.

File: models/dataset.py
Comment (NicolasLegendre1) [reply models/dataset.py:18]: this is the path to data

Comment (fredericpoitevin) [reply models/dataset.py:18]: OK we can fix that later. 

I have two more specific comments:
- It does not appear very clearly why the path is dependent on the availability of torch.cuda
- I doubt the data will be accessible to any user which means most users won't be able to use this as it is

File: models/vae_parameters.json
Comment (fredericpoitevin) [inline models/vae_parameters.json:1]: Nice!

File: models/dataset.py
Comment (fredericpoitevin) [inline models/dataset.py:130]: Much better! Thanks!

**Review by fredericpoitevin — COMMENTED** [review]
It is really starting to look great!
I have only one small reservation about the `CUDA` part but otherwise I'm happy to task the risk to merge despite the lack of unit tests

File: models/dataset.py
Comment (ninamiolane) [inline models/dataset.py:16]: Don't put paths that are user-dependent in the main codebase.

File: models/dataset.py
Comment (ninamiolane) [reply models/dataset.py:29]: Finish cleaning the docstring

File: models/dataset.py
Comment (ninamiolane) [inline models/dataset.py:69]: Finish cleaning docstring

File: models/vae_parameters.json
Comment (ninamiolane) [reply models/vae_parameters.json:1]: 🎉  woohoo!

File: models/vae_parameters.json
Comment (ninamiolane) [inline models/vae_parameters.json:14]: what's the difference between 4points and 4points1? maybe use more descriptive names?

**Review by ninamiolane — COMMENTED** [review]
Muuuuuch better, thank you very much for making the changes! We mostly need to clean the docstrings... and add the unit-tests. Since there are a lot of functions in this PR, this might mean a lot of unit-tests. If you can do all of them at the same time, that is great - otherwise we can close this PR and divide it into a series of smaller PRs. Your call!

File: models/dataset.py
Comment (ninamiolane) [inline models/dataset.py:81]: organize_dataset -> split_dataset

(naming is important, our future users will heavily rely on it to understand what the functions do).

File: models/dataset.py
Comment (ninamiolane) [inline models/dataset.py:108]: Finish cleaning docstring

**Review by ninamiolane — COMMENTED** [review]
Looking good now! Please finish cleaning the docstrings.

Unfortunately, I do not want to merge this without the unit-tests, bcs of the broken window theory:
"The Broken Window Theory suggests that when bad behavior is not corrected immediately, it shows people that there is no downside to breaking the rules, practices or standards. If there is no negative outcome, cutting corners becomes acceptable and in time quality always decreases. The widespread of these problems, especially in software development are clear when you see the estimated costs of bad code or bad development practices amounting to hundreds of billions every year. While the problem might be complex the solutions are simple: take the time to do things right, do code review and fix problems when they are discovered."

Since this PR starts to be very large (if you can avoid large PRs in the future, that would be great, it is hard to review large PR's code in details), it is complicated to add unit-tests all at once.

Let me look now at what could be the best path of actions, and I'll comment again here.

**Comment by ninamiolane** [comment]
Looking better now! Thank you @NicolasLegendre1 for the great work 🎉  Please finish cleaning the docstrings.

Unfortunately, I would rather not merge this without the unit-tests, and I would rather not merge the `dataset.py` module here since it belongs to `ioSPI`. This is because I believe in the broken window theory:
"The Broken Window Theory suggests that when bad behavior is not corrected immediately, it shows people that there is no downside to breaking the rules, practices or standards. If there is no negative outcome, cutting corners becomes acceptable and in time quality always decreases. The widespread of these problems, especially in software development are clear when you see the estimated costs of bad code or bad development practices amounting to hundreds of billions every year. While the problem might be complex the solutions are simple: take the time to do things right, do code review and fix problems when they are discovered."
If we, the first and core contributors to this repo, do not follow our rules in our first PRs, it will become too hard to maintain a clean codebase, and if we do not have a clean codebase we will never scale... Of course, it means that the first PRs will be a real pain for everybody, but it will save us time down the road. We can do it!! 💪 

Additionally, this PR is very large, and it is hard to rigorously review everything. If you can submit small PRs in the future that would be awesome 😜 

I would recommend:
- addressing the code reviews of this PR
- closing this PR
- opening a PR with only: models/nn.py, vae_parameters.json and tests/test_reduceSPI/test_models/test_nn.py
- adding the datasets online somewhere (publicly accessible)
- opening a PR on ioSPI with the datasets utils, downloading from the public access (NOT through the path of a specific user, i.e. not using the path on your computer @NicolasLegendre1 ... because it does not work for the other users - unless they mimic the organization of the files on your computer, but that's a pain).

@fredericpoitevin @NicolasLegendre1 what do you think? 🤗 

I will refactor this repo slightly and set up unittests with `pytest` (see my comment on ioSPI, fred, I believe that we should use pytest instead of `nose2`).

**Comment by geoffwoollard** [comment]
For what it's worth, I am very impressed and encouraged by the high quality standards of software engineering. 

It's encouraging because it means that this is a serious project that will be foundational and last a long time. There is attention to detail and long term vision.

**Comment by fredericpoitevin** [comment]
> Looking better now! Thank you @NicolasLegendre1 for the great work 🎉 Please finish cleaning the docstrings.
> 
> Unfortunately, I would rather not merge this without the unit-tests, and I would rather not merge the `dataset.py` module here since it belongs to `ioSPI`. This is because I believe in the broken window theory:
> "The Broken Window Theory suggests that when bad behavior is not corrected immediately, it shows people that there is no downside to breaking the rules, practices or standards. If there is no negative outcome, cutting corners becomes acceptable and in time quality always decreases. The widespread of these problems, especially in software development are clear when you see the estimated costs of bad code or bad development practices amounting to hundreds of billions every year. While the problem might be complex the solutions are simple: take the time to do things right, do code review and fix problems when they are discovered."
> If we, the first and core contributors to this repo, do not follow our rules in our first PRs, it will become too hard to maintain a clean codebase, and if we do not have a clean codebase we will never scale... Of course, it means that the first PRs will be a real pain for everybody, but it will save us time down the road. We can do it!! 💪
> 
> Additionally, this PR is very large, and it is hard to rigorously review everything. If you can submit small PRs in the future that would be awesome 😜
> 
> I would recommend:
> 
> * addressing the code reviews of this PR
> * closing this PR
> * opening a PR with only: models/nn.py, vae_parameters.json and tests/test_reduceSPI/test_models/test_nn.py
> * adding the datasets online somewhere (publicly accessible)
> * opening a PR on ioSPI with the datasets utils, downloading from the public access (NOT through the path of a specific user, i.e. not using the path on your computer @NicolasLegendre1 ... because it does not work for the other users - unless they mimic the organization of the files on your computer, but that's a pain).
> 
> @fredericpoitevin @NicolasLegendre1 what do you think? 🤗
> 
> I will refactor this repo slightly and set up unittests with `pytest` (see my comment on ioSPI, fred, I believe that we should use pytest instead of `nose2`).

100% agree @ninamiolane !
In terms of setting up unittests (I did see your comment and just switched to pytest), etc, I was planning on setting it up in all compSPI repos once successful with ioSPI. Would that be OK?

**Comment by fredericpoitevin** [comment]
> > Looking better now! Thank you @NicolasLegendre1 for the great work 🎉 Please finish cleaning the docstrings.
> > Unfortunately, I would rather not merge this without the unit-tests, and I would rather not merge the `dataset.py` module here since it belongs to `ioSPI`. This is because I believe in the broken window theory:
> > "The Broken Window Theory suggests that when bad behavior is not corrected immediately, it shows people that there is no downside to breaking the rules, practices or standards. If there is no negative outcome, cutting corners becomes acceptable and in time quality always decreases. The widespread of these problems, especially in software development are clear when you see the estimated costs of bad code or bad development practices amounting to hundreds of billions every year. While the problem might be complex the solutions are simple: take the time to do things right, do code review and fix problems when they are discovered."
> > If we, the first and core contributors to this repo, do not follow our rules in our first PRs, it will become too hard to maintain a clean codebase, and if we do not have a clean codebase we will never scale... Of course, it means that the first PRs will be a real pain for everybody, but it will save us time down the road. We can do it!! 💪
> > Additionally, this PR is very large, and it is hard to rigorously review everything. If you can submit small PRs in the future that would be awesome 😜
> > I would recommend:
> > 
> > * addressing the code reviews of this PR
> > * closing this PR
> > * opening a PR with only: models/nn.py, vae_parameters.json and tests/test_reduceSPI/test_models/test_nn.py
> > * adding the datasets online somewhere (publicly accessible)
> > * opening a PR on ioSPI with the datasets utils, downloading from the public access (NOT through the path of a specific user, i.e. not using the path on your computer @NicolasLegendre1 ... because it does not work for the other users - unless they mimic the organization of the files on your computer, but that's a pain).
> > 
> > @fredericpoitevin @NicolasLegendre1 what do you think? 🤗
> > I will refactor this repo slightly and set up unittests with `pytest` (see my comment on ioSPI, fred, I believe that we should use pytest instead of `nose2`).
> 
> 100% agree @ninamiolane !
> In terms of setting up linting and testing (I did see your comment and just switched to pytest), etc, I was planning on setting it up in all compSPI repos once successful with [ioSPI](https://github.com/compSPI/ioSPI/pull/3). Would that be OK?

Correction: just saw your other PR for linting and testing. Whatever works! :D

**Comment by ninamiolane** [comment]
@fredericpoitevin my PR on linting and testing is currently not working: if you get a working PR first, I'd be very grateful if you could put it here too! 😄

**Comment by fredericpoitevin** [comment]
> @fredericpoitevin my PR on linting and testing is currently not working: if you get a working PR first, I'd be very grateful if you could put it here too! 😄

haha I've been struggling with mine for days! But hopefully almost there. What will help going forward is to find a way to speed the process - would allow to iterate faster. In the case of ioSPI (but probably all other repos soon), the bottleneck is creating the environment. I'll work on the following fix: compSPI/compSPI will have the methods to build environments that can be pulled without building by all other repos. https://github.com/compSPI/ioSPI/issues/4

**Comment by ninamiolane** [comment]
I thought you were caching the environment for this reason? It doesn't work?

**Pull request closed without merging**
